README.md
gyselroth PHP Logger Helper
===========================

## Table of contents

* [Installation](#installation)
* [Log-Wrapper](#log-wrapper)
    + [Usage Example 1: Within Slim PHP application](#usage-example-1-within-slim-php-application)
    + [Usage Example 2: Within Zend Framework 1 application](#usage-example-2-within-zend-framework-1-application)
* [Running Tests](#running-tests)
* [Running phpstan](#running-phpstan)
* [Contributing](#contributing)
* [Author and License](#author-and-license)


Installation
------------

```sh
composer require gyselroth/php-helper-zip
```


### Log-Wrapper

To allow classes of this helper library to log, using the standard logger of the framework used in the rest of 
your application, the library includes ```Gyselroth\Helper\LogWrapper```.
After bootstrapping your application or its service container (if using a framework able of dependency-injection),
the host application's logger can be registered with the Logger-Wrapper (skip this step if you do not use the helper library 
to write any log-entries):

  
#### Usage Example 1: Within Slim PHP application
```php
<?php 
 use Gyselroth\Helper;
 //...
 
 $app = new \Slim\App($settings);

 $container = $app->getContainer();
 
 //...

 $container['logger'] = function (/*...*/) {
     // Callback to PSR-7 logger, e.g. Monolog
     //...
 };
 
 // Register host application's logger component within gyselroth Helper's logger wrapper
 new \Gyselroth\HelperLog\LoggerWrapper($app->getContainer()['logger']);
```


#### Usage Example 2: Within Zend Framework 1 application

As there's no service container for dependency injection within Zend Framework 1 yet,
the helper classes will by convention draw your logger (probably Zend_Log) from the Zend_Registry

```php
<?php 
 use Gyselroth\Helper\;
 //...
 
 $app = new Zend_Application($env, $config);
 $app->bootstrap();
  
 // Register ZF1 logger class within the helper library's logger-wrapper
 new LoggerWrapper('App_Log');
 
 $app->run();
```


Running Tests
-------------

```sh
composer test
```


Running phpstan
---------------

```sh
 vendor/bin/phpstan analyse /srv/www/trunk/src -c /srv/www/trunk/var/ci/phpstan/phpstan.neon
```


Contributing
------------

See [CONTRIBUTING.md](https://github.com/gyselroth/php-helper/blob/master/CONTRIBUTING.md)


History
-------

See [CHANGELOG.md](https://github.com/gyselroth/php-helper/blob/master/CHANGELOG.md)


Author and License
------------------

Copyright 2017-2020 gyselroth™ (http://www.gyselroth.com)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0":http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. 


### Used Open Source Software

Open source packages used by the gyselroth Helper Library are copyright of their vendors, see related licenses within
the vendor packages.


CHANGELOG.md
gyselroth Logger-Wrapper Library
================================

Version 0.2.20 - 2020/09/28
---------------------------
* Initial release after extraction from gyselroth/php-helper


tests/HelperTestCase.php
<?php

/**
 * Copyright (c) 2017-2020 gyselroth™  (http://www.gyselroth.net)
 *
 * @package \gyselroth\Helper
 * @author  gyselroth™  (http://www.gyselroth.com)
 * @link    http://www.gyselroth.com
 * @license Apache-2.0
 */

namespace Tests;

use Gyselroth\HelperLog\LoggerWrapper;
use Monolog\Handler\StreamHandler;
use Monolog\Logger;
use PHPUnit\Framework\TestCase;

class HelperTestCase extends TestCase {
    public static string $rootPath;

    protected string $_pathToLogfile = __DIR__ . '/tmp/app.log';
    protected $_logMock;

    protected ?LoggerWrapper $_logger;

    /**
     * @throws \Exception
     */
    protected function setUp(): void
    {
        self::emptyTempFolder();
        /** @noinspection PhpUnhandledExceptionInspection */
        /** @noinspection PhpParamsInspection */
        $this->_logger = new LoggerWrapper($this->_setUpLogger(), true, '.');
    }

    protected function tearDown(): void
    {
        self::emptyTempFolder();
        if (\is_object($this->_logger)) {
            /** @noinspection PhpUndefinedMethodInspection */
            /** @noinspection ImplicitMagicMethodCallInspection */
            $this->_logger->__destruct();
            $this->_logger = null;
        }
    }

    private function emptyTempFolder(): void
    {
        $tmpPath = self::getGlobalTmpPath();
        if (is_dir($tmpPath)) {
            self::rmdirRecursive($tmpPath);
        }
    }

    /**
     * @param  bool $useStdOut
     * @param  string $logLevel
     * @return Logger
     * @throws \Exception
     */
    private function _setUpLogger($useStdOut = false, $logLevel = 'DEBUG'): Logger
    {
        $path =  $useStdOut ? 'php://stdout' : __DIR__ . '/../var/logs/phpunit.log';

        return (new Logger('phpunit'))
//            ->pushProcessor(new Monolog\Processor\UidProcessor())
            ->pushHandler(new StreamHandler($path, $logLevel));
    }

    /**
     * @param  bool $createIfNotExists
     * @return string
     * @throws \RuntimeException
     */
    public static function getGlobalTmpPath($createIfNotExists = false): string
    {
        $path = self::getRootPath() . '/tmp';
        if ($createIfNotExists
            && !\is_dir($path)
            && !\mkdir($path)
        ) {
            throw new \RuntimeException(\sprintf('Failed create directory "%s"', $path));
        }

        return $path;
    }

    /**
     * @return string
     * @singleton
     */
    private static function getRootPath(): string
    {
        if (empty(self::$rootPath)) {
            /** @noinspection ReturnFalseInspection */
            $pathDelimiter = false !== \strpos(__DIR__, '/vendor/')
                // __DIR__ is e.g. '/srv/www/trunk/vendor/gyselroth/....../HelperFile
                ? '/vendor/'

                // Fallback: helper-package seems to be itself the project at hand
                // (is not installed in one of composer's vendor sub directories)
                // __DIR__ is e.g. '/srv/www/trunk/src/Gyselroth/Helper'
                : '/src/';

            self::$rootPath = \explode($pathDelimiter, __DIR__)[0];
        }

        return self::$rootPath;
    }

    /**
     * @param  string|array $path
     * @return bool
     */
    private static function rmdirRecursive($path): bool
    {
        if (\is_array($path)) {
            foreach ($path as $pathSingle) {
                self::rmdirRecursive($pathSingle);
            }

            return true;
        }

        if (!\is_dir($path)) {
            return false;
        }

        /** @noinspection ReturnFalseInspection */
        $files = \scandir($path);

        if (!$files) {
            return false;
        }

        foreach ($files as $file) {
            if ('.' !== $file
                && '..' !== $file
            ) {
                $rmPath = $path . '/' . $file;

                if (\is_dir($rmPath)) {
                    self::rmdirRecursive($rmPath);
                } else {
                    \unlink($rmPath);
                }
            }
        }

        /** @noinspection ReturnFalseInspection */
        \reset($files);

        return \rmdir($path);
    }
}



tests/LoggerWrapperTest.php
<?php

/**
 * Copyright (c) 2017-2020 gyselroth™  (http://www.gyselroth.net)
 *
 * @package \gyselroth\HelperLog
 * @author  gyselroth™  (http://www.gyselroth.com)
 * @link    http://www.gyselroth.com
 * @license Apache-2.0
 */

namespace Tests;

use Gyselroth\HelperLog\LoggerWrapper;
use Tests\Fixtures\CustomLogMock;
use Tests\Fixtures\MonologMock;
use Tests\Fixtures\TestFilesHelper;


class LoggerWrapperTest extends HelperTestCase
{
    protected string $_pathToLogfile = __DIR__ . '/tmp/app.log';
    protected $_logMock;
    protected ?LoggerWrapper $_logger;

    protected function setUp(): void
    {
        TestFilesHelper::emptyTmpDirectory();
    }

    protected function tearDown(): void
    {
        TestFilesHelper::removeTmpDirectory();
        if (\is_object($this->_logger)) {
            /** @noinspection PhpUndefinedMethodInspection */
            /** @noinspection ImplicitMagicMethodCallInspection */
            $this->_logger->__destruct();
            $this->_logger = null;
        }
    }

//    /**
//     * Test LoggerWrapper::__construct without $loggerReference being an object
//     */
//    public function testLoggerWrapperConstructWithoutLoggerObject()
//    {
//        $this->_logger = new LoggerWrapper('test');
//    }

    /**
     * Test LoggerWrapper::isDevEnvironment without $isDevEnvironment
     * @throws \Gyselroth\HelperLog\Exception\LoggerException
     */
    public function testIsDevEnvironmentDefault(): void
    {
        $this->_logger = new LoggerWrapper(
            $this->createMock(CustomLogMock::class)
        );

        $this->assertFalse($this->_logger->isDevEnvironment());
    }

    /**
     * Test LoggerWrapper::isDevEnvironment with $isDevEnvironment = true
     * @throws \Gyselroth\HelperLog\Exception\LoggerException
     */
    public function testIsDevEnvironmentTrue(): void
    {
        $this->_logger = new LoggerWrapper(
            $this->createMock(CustomLogMock::class),
            true
        );

        $this->assertTrue($this->_logger->isDevEnvironment());
    }

//    /**
//     * Test LoggerWrapper::isDevEnvironment with $isDevEnvironment = false
//     * @throws \Gyselroth\HelperLog\Exception\LoggerException
//     */
//    public function testIsDevEnvironmentFalse()
//    {
//        $this->_logger = new LoggerWrapper($this->createMock(CustomLogMock::class), false);
//        $this->assertFalse($this->_logger::isDevEnvironment());
//    }

    /**
     * Test LoggerWrapper::log default priority (info) to custom log (e.g. Zend_Log)
     * @throws \Exception
     */
    public function testLogInfoToCustomLog(): void
    {
        $this->_logger = $this->_setUpCustomLogger();
        $this->_logger->log('test message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'test message [priority: 6]');
    }

    /**
     * Test LoggerWrapper::log emergency message with log options to custom log
     * @throws \Exception
     */
    public function testLogEmergencyWithOptionsToCustomLog(): void
    {
        $this->_logger = $this->_setUpCustomLogger();
        $this->_logger->log('test message', 'emergency', ['option_1', 'option_2', 'option_3']);

        $this->assertStringEqualsFile(
            $this->_pathToLogfile,
            'test message [priority: 0] [options: option_1, option_2, option_3]'
        );
    }

    /**
     * Test LoggerWrapper::log info to monolog with Streamhandler - no options
     * @throws \Exception
     */
    public function testLogInfoToMonolog(): void
    {
        $this->_logger = $this->_setUpMonoLogger();

        $this->_logger->log('test message', 'info');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'test message [priority: 200]');
    }

    /**
     * Test LoggerWrapper::log critical message to monolog without Streamhandler - no options
     * @throws \Exception
     */
    public function testLogCriticalToMonologWithoutStreamHandler(): void
    {
        $this->_logger = $this->_setUpMonoLogger(false);

        $this->_logger->log('test message', 'critical');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'test message [priority: 500]');
    }

    /**
     * Test LoggerWrapper::alert to custom log
     * @throws \Exception
     */
    public function testAlertToCustomLog(): void
    {
        $this->_logger = $this->_setUpCustomLogger();

        $this->_logger->alert('alert message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'alert message [priority: 1]');
    }

    /**
     * Test LoggerWrapper::alert to mono log
     * @throws \Exception
     */
    public function testAlertToMonoLog(): void
    {
        $this->_logger = $this->_setUpMonoLogger();

        $this->_logger->alert('alert message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'alert message [priority: 550]');
    }

    /**
     * Test LoggerWrapper::crit to custom log
     * @throws \Exception
     */
    public function testCritToCustomLog(): void
    {
        $this->_logger = $this->_setUpCustomLogger();

        $this->_logger->crit('crit message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'crit message [priority: 2]');
    }

    /**
     * Test LoggerWrapper::crit to mono log
     * @throws \Exception
     */
    public function testCritToMonoLog(): void
    {
        $this->_logger = $this->_setUpMonoLogger();

        $this->_logger->crit('crit message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'crit message [priority: 500]');
    }

    /**
     * Test LoggerWrapper::debug to custom log
     * @throws \Exception
     */
    public function testDebugToCustomLog(): void
    {
        $this->_logger = $this->_setUpCustomLogger();

        $this->_logger->debug('debug message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'debug message [priority: 7]');
    }

    /**
     * Test LoggerWrapper::debug to mono log
     * @throws \Exception
     */
    public function testDebugToMonoLog(): void
    {
        $this->_logger = $this->_setUpMonoLogger();

        $this->_logger->debug('debug message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'debug message [priority: 100]');
    }

    /**
     * Test LoggerWrapper::emerg to custom log
     * @throws \Exception
     */
    public function testEmergToCustomLog(): void
    {
        $this->_logger = $this->_setUpCustomLogger();

        $this->_logger->emerg('emerg message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'emerg message [priority: 0]');
    }

    /**
     * Test LoggerWrapper::emerg to mono log
     * @throws \Exception
     */
    public function testEmergToMonoLog(): void
    {
        $this->_logger = $this->_setUpMonoLogger();

        $this->_logger->emerg('emerg message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'emerg message [priority: 600]');
    }

    /**
     * Test LoggerWrapper::error to custom log
     * @throws \Exception
     */
    public function testErrorToCustomLog(): void
    {
        $this->_logger = $this->_setUpCustomLogger();

        $this->_logger->error('error message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'error message [priority: 3]');
    }

    /**
     * Test LoggerWrapper::error to mono log
     * @throws \Exception
     */
    public function testErrorToMonoLog(): void
    {
        $this->_logger = $this->_setUpMonoLogger();

        $this->_logger->error('error message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'error message [priority: 400]');
    }

    /**
     * Test LoggerWrapper::info to custom log
     * @throws \Exception
     */
    public function testInfoToCustomLog(): void
    {
        $this->_logger = $this->_setUpCustomLogger();

        $this->_logger->info('info message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'info message [priority: 6]');
    }

    /**
     * Test LoggerWrapper::info to mono log
     * @throws \Exception
     */
    public function testInfoToMonoLog(): void
    {
        $this->_logger = $this->_setUpMonoLogger();

        $this->_logger->info('info message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'info message [priority: 200]');
    }

    /**
     * Test LoggerWrapper::notice to custom log
     * @throws \Exception
     */
    public function testNoticeToCustomLog(): void
    {
        $this->_logger = $this->_setUpCustomLogger();

        $this->_logger->notice('notice message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'notice message [priority: 5]');
    }

    /**
     * Test LoggerWrapper::notice to mono log
     * @throws \Exception
     */
    public function testNoticeToMonoLog(): void
    {
        $this->_logger = $this->_setUpMonoLogger();

        $this->_logger->notice('notice message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'notice message [priority: 250]');
    }

    /**
     * Test LoggerWrapper::warning to custom log
     * @throws \Exception
     */
    public function testWarningToCustomLog(): void
    {
        $this->_logger = $this->_setUpCustomLogger();

        $this->_logger->warning('warning message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'warning message [priority: 4]');
    }

    /**
     * Test LoggerWrapper::warning to mono log
     * @throws \Exception
     */
    public function testWarningToMonoLog(): void
    {
        $this->_logger = $this->_setUpMonoLogger();

        $this->_logger->warning('warning message');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'warning message [priority: 300]');
    }

    /**
     * Test LoggerWrapper:: logOrDieOnDevOnToCustLogDev on not-dev with log as error = true
     * @throws \Exception
     */
    public function testLogOrDieOnDevToCustLogNotDevLogAsError(): void
    {
        $this->_logger = $this->_setUpCustomLogger();

        $result = $this->_logger->logOrDieOnDev('test message', true, 'result', 'category_1');

        $this->assertStringEqualsFile($this->_pathToLogfile, 'test message [priority: 3] [options: category_1]');
        $this->assertSame($result, 'result');
    }

    /**
     * Test LoggerWrapper:: logOrDieOnDevOnToCustLogDev on not-dev with log as error = false
     * @throws \Exception
     */
    public function testLogOrDieOnDevToCustLogNotDevLogNotAsError(): void
    {
        $this->_logger = $this->_setUpCustomLogger();

        $result = $this->_logger->logOrDieOnDev('test message', false, 'result', 'category_1');

        $this->assertFileNotExists($this->_pathToLogfile);
        $this->assertSame($result, 'result');
    }

//    /**
//     * Test LoggerWrapper:: logOrDieOnDevOnToCustLogDev on Dev => not testable, because script dies!
//     */
//    public function testLogOrDieOnDevOnToCustLogDev()
//    {
//        $this->_logger = $this->_setUpCustomLogger(true);
//        $this->_logger->logOrDieOnDev('test message', true, 'test', 'category_1');
//    }

    /**
     * @param bool $isDevEnvironment
     * @return LoggerWrapper
     * @throws \Gyselroth\HelperLog\Exception\LoggerException
     */
    protected function _setUpCustomLogger(bool $isDevEnvironment = false): LoggerWrapper
    {
        return new LoggerWrapper(
            $this->getMockBuilder(CustomLogMock::class)
                ->setConstructorArgs([$this->_pathToLogfile, $isDevEnvironment])
                ->setMethods(null)
                ->getMock(),
            $isDevEnvironment
        );
    }

    /**
     * @param bool $hasStreamHandlers
     * @return LoggerWrapper
     * @throws \Gyselroth\HelperLog\Exception\LoggerException
     */
    protected function _setUpMonoLogger(bool $hasStreamHandlers = true): LoggerWrapper
    {
        return new LoggerWrapper(
            $this->getMockBuilder(MonologMock::class)
                ->setConstructorArgs([$this->_pathToLogfile])
                ->setMethods(null)
                ->getMock()
        );
    }
}


tests/Fixtures/MonologMock.php
<?php

/**
 * Copyright (c) 2017-2020 gyselroth™  (http://www.gyselroth.net)
 *
 * @package \gyselroth\HelperLog
 * @author  gyselroth™  (http://www.gyselroth.com)
 * @link    http://www.gyselroth.com
 * @license Apache-2.0
 */

namespace Tests\Fixtures;

use \Monolog\Logger;
use \Monolog\Handler\HandlerInterface;

class MonologMock extends Logger
{
    /** @var string */
    private $_pathToLogfile;

    /** @var bool */
    private $_hasStreamHandlers;

    /**
     * Constructor
     *
     * @param string $pathToLogfile
     * @param bool   $hasStreamHandlers
     */
    public function __construct(string $pathToLogfile, bool $hasStreamHandlers = true)
    {
        $this->_pathToLogfile     = $pathToLogfile;
        $this->_hasStreamHandlers = $hasStreamHandlers;
    }

    public function getHandlers(): bool
    {
        return $this->_hasStreamHandlers;
    }

    /**
     * @param  HandlerInterface $handler
     * @return void
     */
    public function pushHandler(HandlerInterface $handler): void
    {
        $this->_hasStreamHandlers = true;
    }

    /**
     * @param  int|string $priority
     * @param  string     $message
     * @param  array      $options
     * @return void
     */
    public function log($priority, $message, array $options = []): void
    {
        $message .= " [priority: $priority]" . (!empty($options) ? ' [options: ' . \implode(', ', $options) . ']' : '');

        /** @noinspection ReturnFalseInspection */
        \file_put_contents($this->_pathToLogfile, $message, FILE_APPEND);
    }
}


src/Gyselroth/HelperLog/LoggerWrapper.php
<?php

/**
 * Copyright (c) 2017-2020 gyselroth™  (http://www.gyselroth.net)
 *
 * @package \gyselroth\HelperLog
 * @author  gyselroth™  (http://www.gyselroth.com)
 * @link    http://www.gyselroth.com
 * @license Apache-2.0
 */

namespace Gyselroth\HelperLog;

use Gyselroth\HelperLog\Exception\LoggerException;
use Monolog\Handler\StreamHandler;

/**
 * Logger Wrapper - Optional service container for using logger component of different application host frameworks
 * Simple service container for DI into static and framework-agnostic helper classes
 *
 * @package Gyselroth\Helper
 */
class LoggerWrapper
{
    // Options items keys
    public const OPT_CATEGORY = 'category';
    public const OPT_PARAMS   = 'parameters';

    /**
     * Alert:     Action must be taken immediately
     * Emergency: System is unusable
     * Notice:    Normal but significant condition
     */
    public const PRIORITY_PSR3_EMERGENCY = 'emergency';
    public const PRIORITY_PSR3_ALERT     = 'alert';
    public const PRIORITY_PSR3_CRITICAL  = 'critical';
    public const PRIORITY_PSR3_ERROR     = 'error';
    public const PRIORITY_PSR3_WARNING   = 'warning';
    public const PRIORITY_PSR3_NOTICE    = 'notice';
    public const PRIORITY_PSR3_INFO  = 'info';
    public const PRIORITY_PSR3_DEBUG = 'debug';

    private const ZF1_PRIORITIES = [
        self::PRIORITY_PSR3_DEBUG     => 7,
        self::PRIORITY_PSR3_INFO      => 6,
        self::PRIORITY_PSR3_NOTICE    => 5,
        self::PRIORITY_PSR3_WARNING   => 4,
        self::PRIORITY_PSR3_ERROR     => 3,
        self::PRIORITY_PSR3_CRITICAL  => 2,
        self::PRIORITY_PSR3_ALERT     => 1,
        self::PRIORITY_PSR3_EMERGENCY => 0
    ];

    private const MONOLOG_LEVELS = [
        self::PRIORITY_PSR3_DEBUG     => 100,
        self::PRIORITY_PSR3_INFO      => 200,
        self::PRIORITY_PSR3_NOTICE    => 250,
        self::PRIORITY_PSR3_WARNING   => 300,
        self::PRIORITY_PSR3_ERROR     => 400,
        self::PRIORITY_PSR3_CRITICAL  => 500,
        self::PRIORITY_PSR3_ALERT     => 550,
        self::PRIORITY_PSR3_EMERGENCY => 600
    ];

    /** @var LoggerWrapper|null */
    protected static $instance;

    protected static string $loggerClassName;

    /** @var \Monolog\Logger */
    protected static $logger;

    protected static bool $isDevEnvironment;

    protected static string $logPath;

    /**
     * Constructor
     *
     * @param  callable|string $loggerReference PSR-7 logger class callback, or logger class name as string
     * @param  bool            $isDevEnvironment
     * @param  string          $logPath
     * @throws LoggerException
     * @singleton
     */
    public function __construct($loggerReference, bool $isDevEnvironment = false, string $logPath = '')
    {
        if (null === self::$instance) {
            if (!\is_object($loggerReference)) {
                throw new LoggerException(
                    'first argument to LoggerWrapper::__construct must be an object (logger)'
                );
            }

            self::$logger           = $loggerReference;
            self::$logPath          = $logPath;
            self::$isDevEnvironment = $isDevEnvironment;
            self::$instance         = $this;
        }
    }

    /**
     * Mute output
     *
     * @throws LoggerException
     */
    public static function mute(): void
    {
        new LoggerWrapper(static function(){});
    }

    /**
     * Destructor: unset instance
     */
    public function __destruct()
    {
        self::$instance = null;
    }

    /**
     * @return bool
     */
    public static function isDevEnvironment(): bool
    {
        return (bool)self::$isDevEnvironment;
    }

    /**
     * Undefined method handler.
     *
     * @param  string $method
     * @param  array  $params
     * @throws LoggerException
     */
    public function __call(string $method, array $params): void
    {
        self::call($method, $params);
    }

    /**
     * Undefined method handler. Implements shortcut methods to log() w/ status -
     * crit(), debug(), error(), info(), warning()
     *
     * @param  string $method
     * @param  array  $params
     * @throws LoggerException
     */
    private static function call(string $method, array $params): void
    {
        $priority = \strtoupper($method);

        if ([] === $params) {
            throw new LoggerException('Missing log message');
        }

        /** @noinspection ReturnNullInspection */
        $message = \array_shift($params);

        /** @noinspection ReturnNullInspection */
        $extras = [] !== $params
            ? \array_shift($params)
            : null;

        self::log($message, $priority, $extras);
    }

    public static function log(string $message, string $priority = self::PRIORITY_PSR3_INFO, array $options = []): void
    {
        /** @noinspection PhpUnnecessaryFullyQualifiedNameInspection */
        if (self::$logger instanceof \Monolog\Logger) {
            // TODO: should logger wrapper really have to care about logfile and loglevel?
            if (empty(self::$logger->getHandlers())) {
                $streamHandler = new StreamHandler(self::$logPath, self::MONOLOG_LEVELS[self::PRIORITY_PSR3_INFO]);

                self::$logger->pushHandler($streamHandler);
            }
            self::$logger->log(self::getMonologLevelByPsr3($priority), $message, $options);
        } else {
            self::$logger->log($message, self::getZf1PriorityByPsr3($priority), $options);
        }
    }

    /**
     * @param  string $message
     * @param  array  $options
     * @throws \Exception
     */
    public static function alert(string $message, array $options = []): void
    {
        if (null !== self::$instance) {
            self::log($message, self::PRIORITY_PSR3_ALERT, $options);
        }
    }

    /**
     * @param  string $message
     * @param  array  $options
     * @throws \Exception
     */
    public static function crit(string $message, array $options = []): void
    {
        if (null !== self::$instance) {
            self::log($message, self::PRIORITY_PSR3_CRITICAL, $options);
        }
    }

    /**
     * @param  string $message
     * @param  array  $options
     * @throws \Exception
     */
    public static function debug(string $message, array $options = []): void
    {
        if (null !== self::$instance) {
            self::log($message, self::PRIORITY_PSR3_DEBUG, $options);
        }
    }

    /**
     * @param  string $message
     * @param  array  $options
     * @throws \Exception
     */
    public static function emerg(string $message, array $options = []): void
    {
        if (null !== self::$instance) {
            self::log($message, self::PRIORITY_PSR3_EMERGENCY, $options);
        }
    }

    /**
     * @param  string $message
     * @param  array  $options
     * @throws \Exception
     */
    public static function error(string $message, array $options = []): void
    {
        if (null !== self::$instance) {
            self::log($message, self::PRIORITY_PSR3_ERROR, $options);
        }
    }

    /**
     * @param  string $message
     * @param  array  $options
     * @throws \Exception
     */
    public static function info(string $message, array $options = []): void
    {
        if (null !== self::$instance) {
            self::log($message, self::PRIORITY_PSR3_INFO, $options);
        }
    }

    /**
     * @param  string $message
     * @param  array  $options
     * @throws \Exception
     */
    public static function notice(string $message, array $options = []): void
    {
        if (null !== self::$instance) {
            self::log($message, self::PRIORITY_PSR3_NOTICE, $options);
        }
    }

    /**
     * @param  string $message
     * @param  array  $options
     * @throws \Exception
     */
    public static function warning(string $message, array $options = []): void
    {
        if (null !== self::$instance) {
            self::log($message, self::PRIORITY_PSR3_WARNING, $options);
        }
    }

    /**
     * @param  string                                     $message
     * @param  bool                                       $logAsError
     * @param  array|bool|int|object|resource|string|null $returnValue
     * @param  string                                     $logCategory
     * @return array|bool|int|object|resource|string|null
     * @throws \Exception
     */
    public static function logOrDieOnDev(
        string $message,
        bool $logAsError = false,
        $returnValue = null,
        string $logCategory = null
    )
    {
        if (self::$isDevEnvironment) {
            die($message . "\n");
        }

        if (false !== $logAsError) {
            self::error($message, [self::OPT_CATEGORY => $logCategory]);
        }

        return $returnValue;
    }

    private static function getZf1PriorityByPsr3(string $priority = ''): int
    {
        return \array_key_exists($priority, self::ZF1_PRIORITIES)
            ? self::ZF1_PRIORITIES[$priority]
            : self::ZF1_PRIORITIES[self::PRIORITY_PSR3_INFO];
    }

    private static function getMonologLevelByPsr3(string $priority = ''): int
    {
        $priority = \strtolower($priority);

        return \array_key_exists($priority, self::MONOLOG_LEVELS)
            ? self::MONOLOG_LEVELS[$priority]
            : self::MONOLOG_LEVELS[self::PRIORITY_PSR3_INFO];
    }
}
